linker/shared/src/main/scala/org/scalajs/linker/checker/IRChecker.scala

```diff
@@ -253,7 +253,37 @@ private final class IRChecker(unit: LinkingUnit, reporter: ErrorReporter) {
               _:JSSuperSelect | _:JSGlobalRef =>
         }
         typecheckExpr(lhs, env)
-        typecheckExpect(rhs, env, lhs.tpe)
+
+        val expectedRhsTpe = lhs match {
+          case ArraySelect(array, _) =>
+            /* Array assignments are unsound due to covariance of arrays
+             * To maintain the subtyping relationship in the IR, we have to
+             * allow assignments of any type to
+             * - Array[Object] (that's expected)
+             * - and its subtypes (that's not expected)
+             */
+
+            array.tpe match {
+              case ArrayType(ArrayTypeRef(PrimRef(tpe), 1), _) => tpe
+
+              case _: ArrayType =>
+                /* Subtype of Array[Object]
+                 * Recall: `Array[Array[A]] <: Array[Object]` even for primitive `A`
+                 * Allow any rhs.
+                 */
+                AnyType
+
+              case _ =>
+                /* Ill typed IR, typecheck of lhs above will emit an error.
```

Comment:
It's not necessarily ill-typed, since having `NullType` (or `NothingType`?) would be valid IR, but we would also need to enter this case and return `AnyType`.